src/EventGridExtension/OutputBinding/EventGridAsyncCollector.cs

```diff
@@ -0,0 +1,53 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.EventGrid;
+using Microsoft.Azure.EventGrid.Models;
+
+namespace Microsoft.Azure.WebJobs.Extensions.EventGrid
+{
+    public sealed class EventGridAsyncCollector : IAsyncCollector<EventGridEvent>
+    {
+        // use IEventGridClient for mocking test
+        private readonly IEventGridClient _client;
+        private readonly IList<EventGridEvent> _eventsToSend = new List<EventGridEvent>();
+        private readonly string _topicHostname;
+
+        private ManualResetEventSlim _canAdd = new ManualResetEventSlim(true);
```

Comment:
Consider using a `SemaphoreSlim` (or even a simple lock here) as it would be a better fit. There are  some scenarios where the current implementation still fails to properly synchronize.

Reply:
can't use `lock` due to the `await` call to `PublishEventsAsync` (can't use await with body of a lock)

will change to SemaphoreSlim

Reply:
That was my point down below. The call to `PublishEventsAsync` doesn't have to occur within the scope of that lock. You can just lock long enough to copy (or replace) the collection and from that point on, you'll no longer be working with a shared resource. That would allow you to use a simple lock (although, the semaphore is still a good option).

The semaphore just makes here as it is easier to reason through in your scenario and would avoid issues like what the current implementation has now, where races are still possible.

Reply:
Can you elaborate on a scenario where a block would still happen, for my better understanding?

Reply:
It would be a rare race (though, which become harder to troubleshoot), but if `AddAsync` and `FlushAsync` were called concurrently, and the `.Wait()` call in `AddAsync` happens right before the `.Reset()` call in `FlushAsync` (which doesn't block waiting for the modification to finish, and that's where the issue comes from), you're now in a race between the list modification and the enumerations done in `FlushAsync`.

Reply:
ah, so we could fix it another way by adding `Wait` in to `FlushAsync` and `Reset/Set` in to `AddAsync`, no?

Reply:
ah, so we could fix it another way by adding `Wait` in to `FlushAsync` and `Reset/Set` in to `AddAsync`, no?

Reply:
That still leaves an opportunity for a race.
Imagine if they're called concurrently, on a signaled event, both calls to `.Wait()` would not block, they would both reset the event and move on.
It also becomes a bit more complex, using a synchronization primitive more appropriate for locking, as opposed to signaling, would simplify this. Essentially you'd have well defined scopes where, while within that scope, you have acquired that lock, exclusively.

Instead of having to make